flowapi/flowapi/permissions.py

```diff
@@ -0,0 +1,356 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+from itertools import takewhile, product, chain
+from typing import List, Iterable, Set, FrozenSet, Tuple, Dict, Any
+
+from flowapi.flowapi_errors import MissingQueryKindError, BadQueryError
+
+
+def get_nested_objects(schema: dict) -> Iterable[Tuple[str, List[str]]]:
+    """
+    Yields tuples of nested objects as the name of the object and a list
+    of the objects it references without the ref prefix.
+
+    Parameters
+    ----------
+    schema : dict
+        Schema to parse
+
+    Yields
+    ------
+    tuple of str, list of str
+
+    """
+    for q, qbod in schema.items():
+        try:
+            refs = [qref["$ref"].split("/")[-1] for qref in qbod["oneOf"]]
+            yield q, refs
+        except (KeyError, TypeError):
+            pass  # Not a oneOf, or a malformed one
+
+
+def get_nested_dict(schema: dict) -> Dict[str, List[str]]:
+    """
+    Gets a dict containing nested objects with the name (sans ref)
+    of the object they reference.
+
+    Parameters
+    ----------
+    schema : dict
+        Schema to parse
+
+    Returns
+    -------
+    dict
+
+    """
+    return dict(get_nested_objects(schema))
+
+
+def get_queries(schema: dict) -> Dict[str, Dict[str, Any]]:
+    """
+    Gets just query objects from a schema.
+
+    Parameters
+    ----------
+    schema : dict
+
+    Returns
+    -------
+    dict
+        Dictionary of query objects
+    """
+    return {
+        q: qbod["properties"]
+        for q, qbod in schema.items()
+        if isinstance(qbod, dict) and "query_kind" in qbod.get("properties", {})
+    }
+
+
+def get_reffed_params(qs: dict, nested: dict) -> dict:
+    """
+    Resolve references in query params.
+
+    Parameters
+    ----------
+    qs : dict
+        All mainline queries
+    nested : dict
+        All nested queries
+    Returns
+    -------
+    dict
+        Queries with any refs to another query substituted for that query
+
+    """
+    q_treed = {}
+    for q, qbod in qs.items():
+        q_treed[q] = {}
+        for param, vals in qbod.items():
+            if "$ref" in vals:
+                ref = vals["$ref"].split("/")[-1]
+                q_treed[q][param] = nested.get(ref, ref)
+    return q_treed
+
+
+def build_tree(roots: List[str], q_treed: dict) -> dict:
```

Comment:
Looks like this doesn't handle multiple layers of nesting. E.g. if `exposed_query_kind` takes a parameter that's either `child_query_kind_a` or `child_query_kind_b`, and `child_query_kind_a` takes a parameter that's either `double_nested_query_kind_1` or `double_nested_query_kind_2`, then the permissions would allow restricting a user to be able to run an `exposed_query_kind` query with `child_query_kind_a` but not `child_query_kind_b`, but wouldn't allow control over which of `double_nested_query_kind_1` or `double_nested_query_kind_2` they could use in `child_query_kind_a`.

I'm not sure whether we'd want to support fine-grained permissions control for multiply-nested queries in this way. In my view it would be nice to be able to do this, but not essential if it would require an over-complicated permissions model (unless we discovered a real situation where this level of control would be advantageous).